File: lollys-lotto-rust-sdk/src/instructions/cranks/crank_lotto_game_winners.rs
use lollys_lotto::state::LottoTicketNumbers;

use crate::instructions::*;
use lollys_lotto::instruction::CrankLottoGameWinners;

pub fn crank_lotto_game_winner(
    round: u64,
    winning_numbers: LottoTicketNumbers,
    winning_numbers_index: [i64; 4],
    authority: &Pubkey,
    user: &Pubkey,
    user_metadata: &Pubkey,
    lotto_game: &Pubkey,
    lotto_game_vault_signer: &Pubkey,
    lotto_game_vault: &Pubkey,
    lotto_ticket: &Pubkey,
    event_emitter: &Pubkey,
) -> Instruction {
    let data = CrankLottoGameWinners {
        round,
        winning_numbers,
        winning_numbers_index,
    }
    .data();
    let accounts = lollys_lotto::accounts::CrankLottoGameWinners {
        authority: *authority,
        lotto_game: *lotto_game,
        lotto_game_vault_signer: *lotto_game_vault_signer,
        lotto_game_vault: *lotto_game_vault,
        user: *user,
        user_metadata: *user_metadata,
        lotto_ticket: *lotto_ticket,
        event_emitter: *event_emitter,
        token_program: token::ID,
    }
    .to_account_metas(None);
    Instruction {
        program_id: lollys_lotto::ID,
        accounts,
        data,
    }
}

File: lollys-lotto-ts-sdk/src/codegen/instructions/crankLottoGameWinners.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CrankLottoGameWinnersArgs {
  round: BN
  winningNumbers: types.LottoTicketNumbersFields
  winningNumbersIndex: Array<BN>
}

export interface CrankLottoGameWinnersAccounts {
  authority: PublicKey
  lottoGame: PublicKey
  lottoGameVaultSigner: PublicKey
  lottoGameVault: PublicKey
  user: PublicKey
  userMetadata: PublicKey
  lottoTicket: PublicKey
  eventEmitter: PublicKey
  tokenProgram: PublicKey
}

export const layout = borsh.struct([
  borsh.u64("round"),
  types.LottoTicketNumbers.layout("winningNumbers"),
  borsh.array(borsh.i64(), 4, "winningNumbersIndex"),
])

export function crankLottoGameWinners(
  args: CrankLottoGameWinnersArgs,
  accounts: CrankLottoGameWinnersAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.authority, isSigner: true, isWritable: false },
    { pubkey: accounts.lottoGame, isSigner: false, isWritable: true },
    {
      pubkey: accounts.lottoGameVaultSigner,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.lottoGameVault, isSigner: false, isWritable: true },
    { pubkey: accounts.user, isSigner: false, isWritable: false },
    { pubkey: accounts.userMetadata, isSigner: false, isWritable: true },
    { pubkey: accounts.lottoTicket, isSigner: false, isWritable: true },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([66, 116, 168, 13, 36, 25, 223, 44])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      round: args.round,
      winningNumbers: types.LottoTicketNumbers.toEncodable(args.winningNumbers),
      winningNumbersIndex: args.winningNumbersIndex,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).subarray(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}
